### PR TITLE
#1908 - E2E Tests - Workflow - Full-time Grants Eligibility

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- E2E tests for full-time eligibility used on the BPMN workflow `fulltime-assessment-*`.
- Fix for BGPD and SBSD that were not working for "BC Private" institutions.
- There is a total of 112 tests executed for the eligibility part. Please take a look at the [PR test execution](https://github.com/bcgov/SIMS/actions/runs/5019161012/jobs/8999352129?pr=1949) to see the full test execution log.
- There is one negative test for the "financial need is at least $1" that is not covered on these tests.
- The tests are organized by the "grant" acronym.
- A helper for students' dependents was created to allow an increase in the family size calculation.
- The tests will be duplicated into the folder `2022-2023` in the next PR to avoid extra work on this one.

image

### CODE REVIEW

File: sources/packages/backend/libs/utilities/src/date-utils.ts [hunk 1]
```diff
@@ -168,6 +168,24 @@ export const addDays = (daysToAdd: number, date?: Date | string): Date => {
     .toDate();
 };
 
+/**
+ * Add units amounts (days, months, years) to a date and returns it as a date only ISO string.
```
Comment (ann-aot) [inline sources/packages/backend/libs/utilities/src/date-utils.ts:172]: not at all a blocker, just my thought, with all amounts (money) in our SIMS, when I first read it, I wondered what is the amount here. Then I realized 😄 the context here.
 can we use something like value or something else here?

File: sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2023-2024.bpmn [hunk 2]
```diff
@@ -1644,7 +1644,7 @@ or (calculatedDataPartnerFedralProvincialPDReceiptCost &gt; 0)</bpmn:conditionEx
     <bpmn:intermediateThrowEvent id="Event_1ki3afc" name="awardEligibilitySBSD">
       <bpmn:extensionElements>
         <zeebe:ioMapping>
-          <zeebe:output source="=if institutionType = &#34;BC Public&#34; or &#34;BC Private&#34; then (&#10;if ((calculatedDataFederalAssessedNeed + calculatedDataProvincialAssessedNeed) &#62;= 1)&#10;  and studentPDStatus = true&#10;then true &#10;else false&#10;)&#10;else&#10;false" target="awardEligibilitySBSD" />
+          <zeebe:output source="=if institutionType = &#34;BC Public&#34; or institutionType = &#34;BC Private&#34; then (&#10;if ((calculatedDataFederalAssessedNeed + calculatedDataProvincialAssessedNeed) &#62;= 1)&#10;  and studentPDStatus = true&#10;then true &#10;else false&#10;)&#10;else&#10;false" target="awardEligibilitySBSD" />
```
Comment (ann-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/fulltime-assessment-2023-2024.bpmn:1647]: 👍

File: sources/packages/backend/workflow/test/2022-2023/fulltime-assessment/fulltime-assessment-2022-2023.e2e-spec.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ import {
 } from "../../test-utils";
 import { PROGRAM_YEAR } from "../constants/program-year.constants";
 
-describe(`E2E Test Workflow fulltime-assessment-${PROGRAM_YEAR}`, () => {
+describe(`E2E Test Workflow fulltime-assessment-${PROGRAM_YEAR}.`, () => {
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/2022-2023/fulltime-assessment/fulltime-assessment-2022-2023.e2e-spec.ts:15]: 👍

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ * @param eligibility eligibility rule to fail.
+ * @returns an not eligible dependent.
+ */
+export function createFakeStudentDependentNotEligible(
+  eligibility: DependentEligibility,
+): StudentDependent {
+  switch (eligibility) {
+    case DependentEligibility.Eligible0To18YearsOld:
+      return {
+        dateOfBirth: addToDateOnlyString(new Date(), -23, "years"),
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:72]: shouldn't it be less than 18 ??

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        DependentEligibility.EligibleOver22YearsOld,
+      ),
+    ];
+    // Act
+    const calculatedAssessment = await executeFulltimeAssessmentForProgramYear(
+      PROGRAM_YEAR,
+      assessmentConsolidatedData,
+    );
+    // Assert
+    // calculatedDataTotalFamilyIncome must be below the threshold for
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:51]: how much is the threshold, just for my info

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts (hunk 4 above)
Comment (ann-aot) [inline sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:72]: shouldn't this be less than -18?

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        dateOfBirth: addToDateOnlyString(new Date(), -22, "years"),
+        attendingPostSecondarySchool: YesNoOptions.No,
+        declaredOnTaxes: YesNoOptions.No,
+      };
+    case DependentEligibility.EligibleOver22YearsOld:
+      return {
+        dateOfBirth: addToDateOnlyString(new Date(), -22.1, "years"),
+        attendingPostSecondarySchool: YesNoOptions.No,
+        declaredOnTaxes: YesNoOptions.No,
+      };
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:93]: so, dependents not attendingPostSecondarySchool and not declaredOnTaxes are ineligible right?

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+} from "../../../test-utils";
+import { PROGRAM_YEAR } from "../../constants/program-year.constants";
+import {
+  DependentEligibility,
+  createFakeStudentDependentEligible,
+  createFakeStudentDependentNotEligible,
+} from "../../../test-utils/factories";
+
+describe(`E2E Test Workflow fulltime-assessment-${PROGRAM_YEAR}-eligibility-CSGD.`, () => {
+  it("Should determine CSGD as eligible when financial need is at least $1 and total family income is under the threshold and eligible dependents are greater than 1.", async () => {
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:13]: so, all these conditions with 1 eligible dependent are not eligible for CSGD?

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      calculatedAssessment.variables.provincialAwardNetCSGDAmount,
+    ).toBeGreaterThan(0);
+  });
+
+  it("Should determine CSGD as not eligible when financial need is at least $1 and total family income is above the threshold and eligible dependents are greater than 1.", async () => {
+    // Arrange
+    const assessmentConsolidatedData =
+      createFakeConsolidatedFulltimeData(PROGRAM_YEAR);
+    assessmentConsolidatedData.studentDataTaxReturnIncome = 100000;
+    // Creates one eligible dependent (between 0-18 years old condition).
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:71]: Line 66 says `eligible dependents are greater than 1`

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    );
+    // Assert
+    expect(calculatedAssessment.variables.awardEligibilityCSGD).toBe(false);
+  });
+
+  it("Should determine CSGD as not eligible when financial need is at least $1 and total family income is above the threshold and eligible dependents is 0.", async () => {
+    // Arrange
+    const assessmentConsolidatedData =
+      createFakeConsolidatedFulltimeData(PROGRAM_YEAR);
+    assessmentConsolidatedData.studentDataTaxReturnIncome = 32999;
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:90]: as per line 17, I think `32999` is under the threshold, but line 86 description does not match

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGP.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    const calculatedAssessment = await executeFulltimeAssessmentForProgramYear(
+      PROGRAM_YEAR,
+      assessmentConsolidatedData,
+    );
+    // Assert
+    // 4000 is the const value set on dmnFullTimeAwardFederalAllowableLimits.limitAwardCSGPAmount.
+    // When the CSGP is eligible its amount will always be 4000. The same value is expected for
+    // provincial and federal grants.
+    expect(calculatedAssessment.variables.federalAwardNetCSGPAmount).toBe(4000);
+    expect(calculatedAssessment.variables.provincialAwardNetCSGPAmount).toBe(
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGP.e2e-spec.ts:23]: shouldn't we also check `expect(calculatedAssessment.variables.awardEligibilityCSGP).toBe(true);`

File: sources/packages/backend/workflow/test/models/assessment.model.ts [hunk 11]
```diff
@@ -144,6 +173,38 @@ export interface CalculatedAssessmentModel {
...
+  federalAwardNetCSGTAmount: number;
+  provincialAwardNetCSGTAmount: number;
+  // BCAG
+  awardEligibilityBCAG: boolean;
+  federalAwardNetBCAGAmount: number;
+  provincialAwardNetBCAGAmount: number;
+  // BCAG2Year
+  awardEligibilityBCAG2Year: number;
+  // BGPD
+  awardEligibilityBGPD: true;
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/models/assessment.model.ts:203]: boolean?

File: sources/packages/backend/workflow/test/models/assessment.model.ts [hunk 12]
```diff
@@ -144,6 +173,38 @@ export interface CalculatedAssessmentModel {
...
+  federalAwardNetBGPDAmount: number;
+  provincialAwardNetBGPDAmount: number;
+  // SBSD
+  awardEligibilitySBSD: true;
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/models/assessment.model.ts:207]: https://github.com/bcgov/SIMS/pull/1949/files#r1199251424

File: sources/packages/backend/workflow/test/test-utils/execute-workflow-instance-utils.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+} from "../models";
+import { PROCESS_INSTANCE_CREATE_TIMEOUT } from "./constants/system-configurations-constants";
+
+/**
+ * Executes the fulltime-assessment-* BPMN workflow.
+ * @param programYear program year to be invoked.
+ * @param assessmentConsolidatedData assessment data.
+ * @returns result of the workflow execution.
+ */
+export async function executeFulltimeAssessmentForProgramYear(
```
Comment (ann-aot) [inline sources/packages/backend/workflow/test/test-utils/execute-workflow-instance-utils.ts:15]: 👍

**Review by ann-aot — COMMENTED** [review]
Good work with the award. now we know the eligibility conditions 👍 Added some comments

File: sources/packages/backend/libs/utilities/src/date-utils.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/utilities/src/date-utils.ts:172]: I see the point. I can change it to value.

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:51]: Thresholds are kept on DMNs and they have all sorts of use. They look like the one below.
image

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:72]: This method is intended to create a dependent that won't be eligible, so the dependent is designed to fail the criteria in different ways for the different eligibilities.

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:93]: - 0-18 years of age: eligible.
- 19-22 years of age: post-sec must be Yes or Tax is Yes.
- > 22 years: eligible if Tax is Yes.

File: sources/packages/backend/workflow/test/models/assessment.model.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/models/assessment.model.ts]: Fixed.

File: sources/packages/backend/workflow/test/models/assessment.model.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/models/assessment.model.ts]: Fixed.

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:90]: Thresholds are different accordingly to the family size.

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:71]: They must be at lest 1, the description is wrong.

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts:13]: CSGD requires at least one eligible dependent. I adjusted the test description.
image

File: sources/packages/backend/libs/utilities/src/date-utils.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/utilities/src/date-utils.ts:172]: Changed!

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGD.e2e-spec.ts]: Fixed.

File: sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGP.e2e-spec.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/test/2023-2024/fulltime-assessment/eligibility/fulltime-assessment-2023-2024-eligibility-CSGP.e2e-spec.ts:23]: Added.

**Review by ann-aot — APPROVED** [review]
Good work with FT grant eligibility. Thanks for doing the changes @andrewsignori-aot 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks great!

File: sources/packages/backend/package.json [hunk 14]
```diff
@@ -27,7 +27,7 @@
     "test:e2e:workers": "npm run migration:run && cross-env ENVIRONMENT=test jest --collect-coverage --verbose --config ./apps/workers/test/jest-e2e.json --forceExit",
     "test:e2e:workers:local": "cross-env ENVIRONMENT=test jest --collect-coverage --verbose --config ./apps/workers/test/jest-e2e.json --forceExit",
     "test:e2e:workflow": "npm run deploy:camunda:definitions && cross-env ENVIRONMENT=test jest --verbose --config ./workflow/test/jest-e2e.json --forceExit",
-    "test:e2e:workflow:local": "cross-env ENVIRONMENT=test jest --verbose --config ./workflow/test/jest-e2e.json --forceExit",
+    "test:e2e:workflow:local": "cross-env ENVIRONMENT=test TZ=UTC jest --verbose --config ./workflow/test/jest-e2e.json --forceExit",
```
Comment (guru-aot) [inline sources/packages/backend/package.json:30]: ❤️

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Pull request merged**